File: backend/app/schemas/club_schema.py
from datetime import date, datetime, time
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Import all club-related enums from the central enums file
from app.schemas.enums import ClubActivityStatus, ClubActivityType, ClubMembershipRole, ClubMembershipStatus, ClubType, MeetingFrequency

# --- Club Schemas ---


class ClubBase(BaseModel):
    name: str = Field(..., max_length=255)
    description: Optional[str] = None
    club_type: ClubType
    logo_url: Optional[str] = Field(None, max_length=500)
    meeting_schedule: Optional[Any] = None  # JSONB field
    meeting_frequency: MeetingFrequency = MeetingFrequency.weekly
    max_members: Optional[int] = Field(None, gt=0)
    registration_open: bool = True
    registration_start_date: Optional[date] = None
    registration_end_date: Optional[date] = None
    club_rules: Optional[str] = None
    objectives: Optional[List[str]] = Field(default_factory=list)  # JSONB as list of strings
    is_active: bool = True


class ClubCreate(ClubBase):
    teacher_in_charge_id: int
    assistant_teacher_id: Optional[int] = None
    academic_year_id: int


class ClubUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    club_type: Optional[ClubType] = None
    logo_url: Optional[str] = Field(None, max_length=500)
    meeting_schedule: Optional[Any] = None
    meeting_frequency: Optional[MeetingFrequency] = None
    max_members: Optional[int] = Field(None, gt=0)
    registration_open: Optional[bool] = None
    registration_start_date: Optional[date] = None
    registration_end_date: Optional[date] = None
    club_rules: Optional[str] = None
    objectives: Optional[List[str]] = None
    is_active: Optional[bool] = None
    teacher_in_charge_id: Optional[int] = None
    assistant_teacher_id: Optional[int] = None


class ClubRead(ClubBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    school_id: int
    teacher_in_charge_id: int
    assistant_teacher_id: Optional[int] = None
    academic_year_id: int
    current_member_count: int
    created_at: datetime
    updated_at: datetime

    # Example of including relational data (you'd need a Pydantic model for Teacher)
    # teacher_in_charge: Optional[Any] = None


# --- Club Membership Schemas ---


class ClubMembershipBase(BaseModel):
    role: ClubMembershipRole = ClubMembershipRole.member
    status: ClubMembershipStatus = ClubMembershipStatus.active
    contribution_score: int = Field(0, ge=0)
    notes: Optional[str] = None


class ClubMembershipCreate(BaseModel):
    student_id: int
    club_id: int
    role: ClubMembershipRole = ClubMembershipRole.member
    status: ClubMembershipStatus = ClubMembershipStatus.active


class ClubMembershipUpdate(BaseModel):
    role: Optional[ClubMembershipRole] = None
    status: Optional[ClubMembershipStatus] = None
    contribution_score: Optional[int] = Field(None, ge=0)
    attendance_count: Optional[int] = Field(None, ge=0)
    notes: Optional[str] = None
    exit_date: Optional[date] = None
    exit_reason: Optional[str] = None


class ClubMembershipRead(ClubMembershipBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    club_id: int
    student_id: int
    approved_by_user_id: UUID
    joined_date: date
    attendance_count: int
    exit_date: Optional[date] = None
    exit_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    # You can add relational models here if needed
    # student: Optional[StudentReadMinimal] = None
    # club: Optional[ClubReadMinimal] = None


# --- Club Activity Schemas ---


class ClubActivityBase(BaseModel):
    activity_name: str = Field(..., max_length=255)
    activity_type: ClubActivityType
    description: Optional[str] = None
    scheduled_date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    venue: Optional[str] = Field(None, max_length=255)
    attendance_mandatory: bool = False
    max_participants: Optional[int] = Field(None, gt=0)
    budget_allocated: Optional[float] = Field(None, ge=0)  # Using float for NUMERIC
    status: ClubActivityStatus = ClubActivityStatus.planned
    outcome_notes: Optional[str] = None
    media_urls: Optional[List[str]] = Field(default_factory=list)


class ClubActivityCreate(ClubActivityBase):
    organized_by_student_id: Optional[int] = None


class ClubActivityUpdate(BaseModel):
    activity_name: Optional[str] = Field(None, max_length=255)
    activity_type: Optional[ClubActivityType] = None
    description: Optional[str] = None
    scheduled_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    venue: Optional[str] = Field(None, max_length=255)
    attendance_mandatory: Optional[bool] = None
    max_participants: Optional[int] = Field(None, gt=0)
    budget_allocated: Optional[float] = None
    status: Optional[ClubActivityStatus] = None
    outcome_notes: Optional[str] = None
    media_urls: Optional[List[str]] = None
    organized_by_student_id: Optional[int] = None


class ClubActivityRead(ClubActivityBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    club_id: int
    organized_by_student_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

File: apps/admin-web/src/app/services/club.schema.ts
// ============================================================================
// CLUB SCHEMAS
// ============================================================================
// Type definitions matching backend endpoints for clubs

export enum ClubType {
  academic = "academic",
  sports = "sports",
  arts = "arts",
  technical = "technical",
  social = "social"
}

export enum MeetingFrequency {
  weekly = "weekly",
  biweekly = "biweekly",
  monthly = "monthly"
}

export enum ClubMembershipRole {
  member = "member",
  secretary = "secretary",
  treasurer = "treasurer",
  president = "president",
  vice_president = "vice_president"
}

export enum ClubMembershipStatus {
  active = "active",
  inactive = "inactive",
  suspended = "suspended",
  alumni = "alumni"
}

export enum ClubActivityType {
  meeting = "meeting",
  workshop = "workshop",
  competition = "competition",
  event = "event",
  project = "project"
}

export enum ClubActivityStatus {
  planned = "planned",
  ongoing = "ongoing",
  completed = "completed",
  cancelled = "cancelled"
}

export interface Club {
  id: number;
  school_id: number;
  academic_year_id: number;
  name: string;
  description?: string;
  club_type: ClubType;
  logo_url?: string;
  meeting_schedule?: any;
  meeting_frequency: MeetingFrequency;
  max_members?: number;
  registration_open: boolean;
  registration_start_date?: string;
  registration_end_date?: string;
  club_rules?: string;
  objectives?: string[];
  teacher_in_charge_id: number;
  assistant_teacher_id?: number;
  current_member_count: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ClubCreate {
  name: string;
  description?: string;
  club_type: ClubType;
  logo_url?: string;
  meeting_schedule?: any;
  meeting_frequency?: MeetingFrequency;
  max_members?: number;
  registration_open?: boolean;
  registration_start_date?: string;
  registration_end_date?: string;
  club_rules?: string;
  objectives?: string[];
  teacher_in_charge_id: number;
  assistant_teacher_id?: number;
  academic_year_id: number;
}

export interface ClubUpdate {
  name?: string;
  description?: string;
  club_type?: ClubType;
  logo_url?: string;
  meeting_schedule?: any;
  meeting_frequency?: MeetingFrequency;
  max_members?: number;
  registration_open?: boolean;
  registration_start_date?: string;
  registration_end_date?: string;
  club_rules?: string;
  objectives?: string[];
  is_active?: boolean;
  teacher_in_charge_id?: number;
  assistant_teacher_id?: number;
}

export interface ClubMembership {
  id: number;
  club_id: number;
  student_id: number;
  role: ClubMembershipRole;
  status: ClubMembershipStatus;
  contribution_score: number;
  notes?: string;
  approved_by_user_id: string;
  joined_date: string;
  attendance_count: number;
  exit_date?: string;
  exit_reason?: string;
  created_at: string;
  updated_at: string;
}

export interface ClubMembershipCreate {
  student_id: number;
  club_id: number;
  role?: ClubMembershipRole;
  status?: ClubMembershipStatus;
}

export interface ClubMembershipUpdate {
  role?: ClubMembershipRole;
  status?: ClubMembershipStatus;
  contribution_score?: number;
  attendance_count?: number;
  notes?: string;
  exit_date?: string;
  exit_reason?: string;
}

export interface ClubActivity {
  id: number;
  club_id: number;
  activity_name: string;
  activity_type: ClubActivityType;
  description?: string;
  scheduled_date: string;
  start_time?: string;
  end_time?: string;
  venue?: string;
  attendance_mandatory: boolean;
  max_participants?: number;
  budget_allocated?: number;
  status: ClubActivityStatus;
  outcome_notes?: string;
  media_urls?: string[];
  organized_by_student_id?: number;
  created_at: string;
  updated_at: string;
}

export interface ClubActivityCreate {
  activity_name: string;
  activity_type: ClubActivityType;
  description?: string;
  scheduled_date: string;
  start_time?: string;
  end_time?: string;
  venue?: string;
  attendance_mandatory?: boolean;
  max_participants?: number;
  budget_allocated?: number;
  status?: ClubActivityStatus;
  outcome_notes?: string;
  media_urls?: string[];
  organized_by_student_id?: number;
}

export interface ClubActivityUpdate {
  activity_name?: string;
  activity_type?: ClubActivityType;
  description?: string;
  scheduled_date?: string;
  start_time?: string;
  end_time?: string;
  venue?: string;
  attendance_mandatory?: boolean;
  max_participants?: number;
  budget_allocated?: number;
  status?: ClubActivityStatus;
  outcome_notes?: string;
  media_urls?: string[];
  organized_by_student_id?: number;
}

export interface ClubKpi {
  active_clubs: number;
  total_members: number;
  events_this_month: number;
  avg_members_per_club: number;
  most_active_club: string;
}
